File: app/src/main/java/com/mxt/anitrend/presenter/base/BasePresenter.java
package com.mxt.anitrend.presenter.base;

import android.content.Context;
import android.content.Intent;
import android.support.v4.app.FragmentActivity;
import android.util.Log;

import com.annimon.stream.Stream;
import com.mxt.anitrend.base.custom.async.WebTokenRequest;
import com.mxt.anitrend.base.custom.presenter.CommonPresenter;
import com.mxt.anitrend.base.interfaces.dao.BoxQuery;
import com.mxt.anitrend.model.entity.anilist.UserStats;
import com.mxt.anitrend.model.entity.anilist.meta.FormatStats;
import com.mxt.anitrend.model.entity.anilist.meta.GenreStats;
import com.mxt.anitrend.model.entity.base.UserBase;
import com.mxt.anitrend.model.entity.crunchy.MediaContent;
import com.mxt.anitrend.model.entity.crunchy.Thumbnail;
import com.mxt.anitrend.service.TagGenreService;
import com.mxt.anitrend.util.CompatUtil;

import java.util.List;
import java.util.Locale;
import java.util.concurrent.TimeUnit;

/**
 * Created by max on 2017/09/16.
 * General presenter for most objects
 */

public class BasePresenter extends CommonPresenter {

    private List<String> favouriteGenres, favouriteTags, favouriteYears, favouriteFormats;

    public BasePresenter(Context context) {
        super(context);
    }

    public final void checkGenresAndTags(FragmentActivity fragmentActivity) {
        Intent intent = new Intent(fragmentActivity, TagGenreService.class);
        fragmentActivity.startService(intent);
    }

    public String getThumbnail(List<Thumbnail> thumbnails) {
        if(CompatUtil.INSTANCE.isEmpty(thumbnails))
            return null;
        return thumbnails.get(0).getUrl();
    }

    public String getDuration(MediaContent mediaContent) {
        if(mediaContent.getDuration() != null) {
            long timeSpan = Integer.valueOf(mediaContent.getDuration());
            long minutes = TimeUnit.SECONDS.toMinutes(timeSpan);
            long seconds = timeSpan - TimeUnit.MINUTES.toSeconds(minutes);
            return (String.format(Locale.getDefault(), seconds < 10 ? "%d:0%d":"%d:%d", minutes, seconds));
        }
        return "00:00";
    }

    public List<String> getTopFavouriteGenres(int limit) {
        if(CompatUtil.INSTANCE.isEmpty(favouriteGenres)) {
            UserStats userStats;
            if (getDatabase().getCurrentUser() != null && (userStats = getDatabase().getCurrentUser().getStats()) != null) {
                if (!CompatUtil.INSTANCE.isEmpty(userStats.getFavouredGenres())) {
                    favouriteGenres = Stream.of(userStats.getFavouredGenres())
                            .sortBy(genreStat -> - genreStat.getAmount())
                            .map(GenreStats::getGenre)
                            .limit(limit).toList();

                }
            }
        }
        return favouriteGenres;
    }

    public List<String> getTopFavouriteTags(int limit) {
        if(CompatUtil.INSTANCE.isEmpty(favouriteTags)) {
            UserStats userStats;
            if (getDatabase().getCurrentUser() != null && (userStats = getDatabase().getCurrentUser().getStats()) != null) {
                if (!CompatUtil.INSTANCE.isEmpty(userStats.getFavouredTags())) {
                    favouriteTags = Stream.of(userStats.getFavouredTags())
                            .sortBy(mediaTagStats -> - mediaTagStats.getAmount())
                            .map(s -> s.getTag().getName())
                            .limit(limit).toList();

                }
            }
        }
        return favouriteTags;
    }

    public List<String> getTopFavouriteYears(int limit) {
        if(CompatUtil.INSTANCE.isEmpty(favouriteYears)) {
            UserStats userStats;
            if (getDatabase().getCurrentUser() != null && (userStats = getDatabase().getCurrentUser().getStats()) != null) {
                if (!CompatUtil.INSTANCE.isEmpty(userStats.getFavouredTags())) {
                    favouriteYears = Stream.of(userStats.getFavouredYears())
                            .sortBy(yearStats -> - yearStats.getAmount())
                            .map(y -> String.valueOf(y.getYear()))
                            .limit(limit).toList();

                }
            }
        }
        return favouriteTags;
    }

    public List<String> getTopFormats(int limit) {
        if(CompatUtil.INSTANCE.isEmpty(favouriteFormats)) {
            UserStats userStats;
            if (getDatabase().getCurrentUser() != null && (userStats = getDatabase().getCurrentUser().getStats()) != null) {
                if (!CompatUtil.INSTANCE.isEmpty(userStats.getFavouredFormats())) {
                    favouriteFormats = Stream.of(userStats.getFavouredFormats())
                            .sortBy(formatStats -> - formatStats.getAmount())
                            .map(FormatStats::getFormat)
                            .limit(limit).toList();

                }
            }
        }
        return favouriteFormats;
    }

    public boolean isCurrentUser(long userId) {
        return getApplicationPref().isAuthenticated() && getDatabase().getCurrentUser() != null &&
                userId != 0 && getDatabase().getCurrentUser().getId() == userId;
    }

    public boolean isCurrentUser(String userName) {
        return getApplicationPref().isAuthenticated() && getDatabase().getCurrentUser() != null &&
                userName != null && getDatabase().getCurrentUser().getName().equals(userName);
    }

    public boolean isCurrentUser(long userId, String userName) {
        if (userName != null)
            return isCurrentUser(userName);
        return isCurrentUser(userId);
    }

    public boolean isCurrentUser(UserBase userBase) {
        return userBase != null && isCurrentUser(userBase.getId());
    }

    public void checkValidAuth() {
        if(getApplicationPref().isAuthenticated()) {
            BoxQuery boxQuery = getDatabase();
            if(boxQuery.getCurrentUser() == null) {
                Log.e("checkValidAuth", "Last attempt to authenticate failed, refreshing session!");
                WebTokenRequest.invalidateInstance(getContext());
            }
        }
    }
}

File: app/src/main/java/com/mxt/anitrend/presenter/base/BasePresenter.kt
package com.mxt.anitrend.presenter.base

import android.content.Context
import android.content.Intent
import androidx.fragment.app.FragmentActivity
import com.annimon.stream.Stream
import com.mxt.anitrend.base.custom.async.WebTokenRequest
import com.mxt.anitrend.base.custom.presenter.CommonPresenter
import com.mxt.anitrend.model.entity.anilist.user.UserStatisticTypes
import com.mxt.anitrend.model.entity.base.UserBase
import com.mxt.anitrend.model.entity.crunchy.MediaContent
import com.mxt.anitrend.model.entity.crunchy.Thumbnail
import com.mxt.anitrend.service.TagGenreService
import com.mxt.anitrend.util.CompatUtil
import com.mxt.anitrend.util.migration.MigrationUtil
import com.mxt.anitrend.util.migration.Migrations
import timber.log.Timber
import java.util.*
import java.util.concurrent.TimeUnit

/**
 * Created by max on 2017/09/16.
 * General presenter for most objects
 */

open class BasePresenter(context: Context?) : CommonPresenter(context) {

    private var favouriteGenres: List<String>? = null
    private var favouriteTags: List<String>? = null
    private var favouriteYears: List<String>? = null
    private var favouriteFormats: List<String>? = null

    fun checkIfMigrationIsNeeded(): Boolean {
        if (!settings.isFreshInstall) {
            val migrationUtil = MigrationUtil.Builder()
                    .addMigration(Migrations.MIGRATION_101_108)
                    .addMigration(Migrations.MIGRATION_109_134)
                    .addMigration(Migrations.MIGRATION_135_136)
                    .build()
            return migrationUtil.applyMigration()
        }
        return true
    }

    fun checkGenresAndTags(fragmentActivity: FragmentActivity) {
        val intent = Intent(fragmentActivity, TagGenreService::class.java)
        fragmentActivity.startService(intent)
    }

    fun getThumbnail(thumbnails: List<Thumbnail>): String? {
        return if (CompatUtil.isEmpty(thumbnails)) null else thumbnails[0].url
    }

    fun getDuration(mediaContent: MediaContent): String {
        if (mediaContent.duration != null) {
            val timeSpan = Integer.valueOf(mediaContent.duration).toLong()
            val minutes = TimeUnit.SECONDS.toMinutes(timeSpan)
            val seconds = timeSpan - TimeUnit.MINUTES.toSeconds(minutes)
            return String.format(Locale.getDefault(), if (seconds < 10) "%d:0%d" else "%d:%d", minutes, seconds)
        }
        return "00:00"
    }

    fun getTopFavouriteGenres(limit: Int): List<String>? {
        if (CompatUtil.isEmpty(favouriteGenres)) {
            val userStats: UserStatisticTypes? = database.currentUser?.statistics
            if (database.currentUser != null && userStats != null) {
                if (!userStats.anime.genres.isNullOrEmpty()) {
                    favouriteGenres = userStats.anime.genres
                            .sortedByDescending {
                                it.count
                            }.filter {
                                it.genre != null
                            }.map {
                                it.genre!!
                            }.take(limit)
                }
            }
        }
        return favouriteGenres
    }

    fun getTopFavouriteTags(limit: Int): List<String>? {
        if (CompatUtil.isEmpty(favouriteTags)) {
            val userStats: UserStatisticTypes? = database.currentUser?.statistics
            if (database.currentUser != null && userStats != null) {
                if (!userStats.anime.tags.isNullOrEmpty()) {
                    favouriteTags = Stream.of(userStats.anime.tags)
                            .sortBy { (_, _, count) -> -count }
                            .filter { (tag) -> tag != null }
                            .map { (tag) -> tag!!.name }
                            .limit(limit.toLong()).toList()
                    favouriteTags = userStats.anime.tags.sortedByDescending {
                        it.count
                    }.filter {
                        it.tag != null
                    }.map {
                        it.tag!!.name
                    }.take(limit)
                }
            }
        }
        return favouriteTags
    }

    fun getTopFavouriteYears(limit: Int): List<String>? {
        if (CompatUtil.isEmpty(favouriteYears)) {
            val userStats: UserStatisticTypes? = database.currentUser?.statistics
            if (database.currentUser != null && userStats != null) {
                if (!userStats.anime.releaseYears.isNullOrEmpty()) {
                    favouriteYears = userStats.anime.releaseYears
                            .sortedByDescending {
                                it.count
                            }.filter {
                                it.releaseYear != null
                            }.map {
                                it.releaseYear?.toString()!!
                            }.take(limit)
                }
            }
        }
        return favouriteTags
    }

    fun getTopFormats(limit: Int): List<String>? {
        if (CompatUtil.isEmpty(favouriteFormats)) {
            val userStats: UserStatisticTypes? = database.currentUser?.statistics
            if (database.currentUser != null && userStats != null) {
                if (!userStats.anime.formats.isNullOrEmpty()) {
                    favouriteFormats = userStats.anime.formats
                            .sortedByDescending {
                                it.count
                            }.filter {
                                it.format != null
                            }.map {
                                it.format!!
                            }.take(limit)
                }
            }
        }
        return favouriteFormats
    }

    fun isCurrentUser(userId: Long): Boolean {
        return settings.isAuthenticated && database.currentUser != null &&
                userId != 0L && database.currentUser?.id == userId
    }

    fun isCurrentUser(userName: String?): Boolean {
        return settings.isAuthenticated && database.currentUser != null &&
                userName != null && database.currentUser?.name == userName
    }

    fun isCurrentUser(userId: Long, userName: String?): Boolean {
        return userName?.let { isCurrentUser(it) } ?: isCurrentUser(userId)
    }

    fun isCurrentUser(userBase: UserBase?): Boolean {
        return userBase != null && isCurrentUser(userBase.id)
    }

    fun checkValidAuth() {
        if (settings.isAuthenticated) {
            if (database.currentUser == null) {
                Timber.tag(TAG).w("Last attempt to authenticate failed, refreshing session!")
                WebTokenRequest.invalidateInstance(context)
            }
        }
    }

    companion object {
        private val TAG = BasePresenter::class.java.simpleName
    }
}
